src/backend/distributed/utils/resource_lock.c

```diff
@@ -690,8 +690,9 @@ SerializeNonCommutativeWrites(List *shardIntervalList, LOCKMODE lockMode)
 {
 	ShardInterval *firstShardInterval = (ShardInterval *) linitial(shardIntervalList);
 	int64 firstShardId = firstShardInterval->shardId;
+	Oid relationId = firstShardInterval->relationId;
 
-	if (ReferenceTableShardId(firstShardId))
+	if (ReferenceTableShardId(firstShardId) || !SingleReplicatedTable(relationId))
```

Comment:
What is the reason for not supporting foreign keys between replication > 1 at the moment?

Reply:
I think the two reasons used to be 1) shardstate needs to cascade in case of failed modification, 2) shard resource locks needs to "cascade" to a common lock.

Since 1) no longer applies, we can maybe allow it?

Reply:
let's keep such ideas on an issue before we forget :) https://github.com/citusdata/citus/issues/5450